lib/core/macros.nim

```diff
@@ -1344,6 +1348,93 @@ proc copy*(node: NimNode): NimNode {.compileTime.} =
+    # `genAst` argument list. The default is unset, to avoid surprising hijacking
+    # of local symbols by symbols in caller scope.
+
+macro genAst*(options: static set[GenAstOpt] = {}, args: varargs[untyped]): untyped =
+  ## Accepts a list of captured `variables = value` and a block and returns the
+  ## AST that represents it. Local `{.inject.}` symbols are captured (eg
+  ## local procs) unless `kNoExposeLocalInjects in options`; additional variables
+  ## are captured as subsequent parameters.
+  runnableExamples:
+    type Foo = enum kfoo0, kfoo1, kfoo2, kfoo3, kfoo4
+
+    macro bar1(x0: static Foo, x1: Foo, x2: Foo, xignored: Foo): untyped =
+      let s0 = "not captured!" ## does not override `s0` from caller scope
+      let s1 = "not captured!" ## does not override `s1=2`
+      let xignoredLocal = kfoo4
+      proc localExposed(): auto = kfoo4 # implicitly captured
+      let x3 = newLit kfoo4
+      result = genAst({}, s1=2, s2="asdf", x0=newLit x0, x1=x1, x2, x3):
+        # echo xignored # would give: Error: undeclared identifier
+        # echo s0 # would give: Error: internal error: expr: var not init s0_237159
+        (s1, s2, x0, x1, x2, x3, localExposed())
+
+    macro bar2(x0: static Foo, x1: Foo, x2: Foo, xignored: Foo): untyped =
+      let s0 = "not captured!" ## does not override `s0` from caller scope
+      let s1 = "not captured!" ## does not override `s1=2`
+      let xignoredLocal = kfoo4
+      let x3 = newLit kfoo4
+      result = genAst({kNoExposeLocalInjects}, s1=2, s2="asdf", x0=newLit x0, x1=x1, x2, x3):
+        ## only captures variables from `genAst` argument list
+        ## uncaptured variables will be set from caller scope (Eg `s0`)
+        ## `x2` is shortcut for the common `x2=x2`
+        doAssert not declared(xignored)      # not in param list!
+        doAssert not declared(xignoredLocal) # ditto
+        (s1, s2, s0, x0, x1, x2, x3)
+
+    block:
+      let s0 = "caller scope!"
+      doAssert bar1(kfoo1, kfoo2, kfoo3, kfoo4) ==
+        (2, "asdf", kfoo1, kfoo2, kfoo3, kfoo4, kfoo4)
+
+    block:
+      let s0 = "caller scope!"
+      doAssert bar2(kfoo1, kfoo2, kfoo3, kfoo4) ==
+        (2, "asdf", "caller scope!", kfoo1, kfoo2, kfoo3, kfoo4)
+
+  let params = newTree(nnkFormalParams, newEmptyNode())
+  let pragmas =
+    if kNoExposeLocalInjects in options:
+      nnkPragma.newTree(ident"dirty")
+    else:
+      newEmptyNode()
+
+  # using `_` as workaround, see https://github.com/nim-lang/Nim/issues/2465#issuecomment-511076669
+  let name = genSym(nskTemplate, "_fun")
+  let call = newCall(name)
+  for a in args[0..^2]:
+    var varName: NimNode
+    var varVal: NimNode
+    case a.kind
+    of nnkExprEqExpr:
+      varName = a[0]
+      varVal = a[1]
+    of nnkIdent:
+      varName = a
+      varVal = a
+    else: error("invalid argument kind: " & $a.kind, a)
+
+    params.add newTree(nnkIdentDefs, [varName, newEmptyNode(), newEmptyNode()])
```

Comment:
NewTree takes ``varargs[NimNode]``, not ``openArray[NimNode]``. I am afraid, but scheduled fixes for varargs will trigger a compiler error here in the future, because ``[varName, newEmptyNode(), newEmptyNode()]`` is not a ``NimNode``, but an array of ``NimNode``.

Reply:
fixed in latest commit